charmhelpers/contrib/network/ovs/__init__.py

```diff
@@ -303,3 +306,158 @@ def port_to_br(port):
         ).decode('UTF-8').strip()
     except subprocess.CalledProcessError:
         return None
+
+
+def _dict_to_vsctl(data, table, entity):
+    """
+    :param data: Additional data to attach to interface
+        The keys in the data dictionary map directly to column names in the
+        OpenvSwitch specified table as defined in DB-SCHEMA [0] referenced in
+        RFC 7047 [1]
+
+        There are some established conventions for keys in the external-ids
+        column of various tables, consult the OVS Integration Guide [2] for
+        more details.
+
+        NOTE(fnordahl): Technically the ``external-ids`` column is called
+        ``external_ids`` (with an underscore) and we rely on ``ovs-vsctl``'s
+        behaviour of transforming dashes to underscores for us [3] so we can
+        have a more pleasant data structure.
+
+        0: http://www.openvswitch.org/ovs-vswitchd.conf.db.5.pdf
+        1: https://tools.ietf.org/html/rfc7047
+        2: http://docs.openvswitch.org/en/latest/topics/integration/
+        3: https://github.com/openvswitch/ovs/blob/
+               20dac08fdcce4b7fda1d07add3b346aa9751cfbc/
+                   lib/db-ctl-base.c#L189-L215
+    :type data: Optional[Dict[str,Union[str,Dict[str,str]]]]
+    :param table: Name of table to operate on
+    :type table: str
+    :param entity: Name of entity to operate on
+    :type entity: str
+    :returns: '--' separated ``ovs-vsctl set`` commands
+    :rtype: Iterator[str]
+    """
+    for (k, v) in data.items():
+        if isinstance(v, dict):
+            entries = {
+                '{}:{}'.format(k, dk): dv for (dk, dv) in v.items()}
+        else:
+            entries = {k: v}
+        for (colk, colv) in entries.items():
+            yield ('--', 'set', table, entity, '{}={}'.format(colk, colv))
+
+
+def add_br(bridge, brdata=None, exclusive=False):
```

Comment:
FYI: I'm preparing a pull-request on top of yours in which I add here a few more args that I need, and especially in which I mark add_bridge() as deprecated in favor of add_br() and add_bridge_port() in favor of add_port().

Do you think this should be done in your pull-request already? If not, then this file is going to be confusing right after your merge (it'll be hard for the consumer to chose between these similar functions)

Reply:
I deliberately kept both as the existing functions do extra work not strictly confined to adding bridges and ports and we need to work out how to handle the transition, so I suggest we handle that in a follow-up.

Reply:
Software engineering issues: cache invalidation, naming things and off-by-one errors, etc.  Can I just point out that `add_bridge` is a MUCH, MUCH better name than `add_br` and you should strive to keep the former rather than the latter.

Reply:
@ajkavanagh The "problem" is that we need to give the new one with new signature a new name and keep the old one for a while, while deprecating it, so we remain backward compatible for a while. Or should we just not be backward compatible?

Reply:
What does the old `add_bridge` do differently to the new `add_br` that requires deprecation and the introduction of a new function?

Reply:
`add_brige` takes a `datapath_type` optional argument, that neither OVN nor neutron-gateway need, but another charm somewhere out there might be using it. But true, we could take this arg as well in the signature of `add_br` and name it `add_bridge`

Reply:
OK, I guess merging the functions now it is then.

Reply:
I guess what is at issue here, is that we have several similar functions that do functionally similar things being merged into library and we need to be happy that we're not duplicating things or making the user experience worse.  Now it may be *correct* to actually deprecate and go with the 'worse' name as a) the world is not perfect, and b) thus where the library started from doesn't allow for a better name at this point.

My questions may be covering old ground, so if that's the case, I'll go away :)

Reply:
No, it's fine and your input is very valuable and needed. You're holding the bar high while I'm suffering from tunnel vision having fixed my eyes on one point in the distance for too long and trying to rushing through to get there.

Having multiple functions which serves more or less the same purpose when not needed is just adding to the pile of technical debt.

As both you and @AurelienLourot has pointed out the wanted functionality can be added to the existing functions without breaking the existing API, so doing that is the right thing to do.